src/main/java/randoop/util/Randomness.java

```diff
@@ -160,71 +176,88 @@ public static int nextRandomInt(int i) {
-    double[] cumulativeWeights = new double[list.size() + 1];
-    cumulativeWeights[0] = 0.0;
     for (int i = 0; i < list.size(); i++) {
       T elt = list.get(i);
-      Double weightOrNull = weights.get(elt);
-      double weight;
-      if (weightOrNull != null) {
-        weight = weightOrNull;
-      } else {
-        weight = elt.getWeight();
-        Log.logPrintf(
-            "randoop.util.Randomness: key %s not found; using intrinsic weight %d%n", elt, weight);
-      }
-      if (weight <= 0) {
+      double weight = getWeight(elt, weights);
+      if (weight < 0) {
         throw new BugInRandoopException("Weight should be positive: " + weight);
       }
-      cumulativeWeights[i] = totalWeight;
       totalWeight += weight;
     }
-    cumulativeWeights[list.size()] = totalWeight;
+
+    return randomMemberWeighted(list, weights, totalWeight);
+  }
+
+  /**
+   * Randomly selects an element from a weighted distribution of elements. These weights are with
+   * respect to each other. They are not normalized (they might add up to any value).
+   *
+   * @param list the list of elements to select from
+   * @param weights the map of elements to their weights; uses the intrinsic weight if the element
+   *     is not a key in the map. Each element's weight must be non-negative. An element with a
+   *     weight of zero will never be selected.
+   * @param <T> the type of the elements in the list
+   * @return a randomly selected element from {@code list}
+   */
+  public static <T extends WeightedElement> T randomMemberWeighted(
+      SimpleList<T> list, Map<T, Double> weights, double totalWeight) {
 
     // Select a random point in interval and find its corresponding element.
-    incrementCallsToRandom("randomMemberWeighted(SimpleList, Map)");
+    incrementCallsToRandom("randomMemberWeighted(SimpleList)");
     double chosenPoint = Randomness.random.nextDouble() * totalWeight;
     if (GenInputsAbstract.selection_log != null) {
       try {
-        GenInputsAbstract.selection_log.write(
-            String.format(
-                "chosenPoint = %s, cumulativeWeights = %s%n", chosenPoint, cumulativeWeights));
+        GenInputsAbstract.selection_log.write(String.format("chosenPoint = %s%n", chosenPoint));
       } catch (IOException e) {
         throw new Error("Problem writing to selection-log", e);
       }
     }
 
-    int index = binarySearchForIndex(cumulativeWeights, chosenPoint);
-    if (GenInputsAbstract.selection_log != null) { // body is expensive
-      logSelection(
-          index,
-          "randomMemberWeighted(List,Map)",
-          String.format(
-              "%n << %s%n    (class %s),%n    %s%n    (class %s, size %s)>>",
-              list, list.getClass(), weights, weights.getClass(), weights.size()));
+    double currentPoint = 0;
+    for (int i = 0; i < list.size(); i++) {
+      currentPoint += getWeight(list.get(i), weights);
+      if (currentPoint >= chosenPoint) {
```

Comment:
Hi Mike,
I'm wondering about the case where we have elements that have weights of value zero.
I was thinking that Randomness should never select an element that has a weight of zero? However, if the list's first element were to have a weight of zero and chosenPoint were to be zero, this code would select that element.

Reply:
This is possible.  There is an approximately 1/2^64 chance of this happening.
We could change `>=` to `<` (which, it occurs to me, might be the correct algorithm anyway), or add an explicit check for two zeroes, or ignore the situation.